src/catalog/src/kvbackend/manager.rs

```diff
@@ -303,36 +301,70 @@ impl CatalogManager for KvBackendCatalogManager {
+        let metadata_manager = self.table_metadata_manager.clone();
+        let catalog = catalog.to_string();
+        let schema = schema.to_string();
+        tokio::task::spawn(async move {
+            let table_id_stream = metadata_manager
+                .table_name_manager()
+                .tables(&catalog, &schema)
+                .map_ok(|(_, v)| v.table_id());
             // Split table ids into chunks
             let mut table_id_chunks = table_id_stream.ready_chunks(BATCH_SIZE);
 
+            let mut task_batch = Vec::with_capacity(CONCURRENCY);
             while let Some(table_ids) = table_id_chunks.next().await {
-                let table_ids = table_ids
+                let table_ids = match table_ids
                     .into_iter()
                     .collect::<std::result::Result<Vec<_>, _>>()
                     .map_err(BoxedError::new)
-                    .context(ListTablesSnafu { catalog, schema })?;
-
-                let table_info_values = self
-                    .table_metadata_manager
-                    .table_info_manager()
-                    .batch_get(&table_ids)
-                    .await
-                    .context(TableMetadataManagerSnafu)?;
-
-                for table_info_value in table_info_values.into_values() {
-                    yield build_table(table_info_value)?;
+                    .context(ListTablesSnafu {
+                        catalog: &catalog,
+                        schema: &schema,
+                    }) {
+                    Ok(table_ids) => table_ids,
+                    Err(e) => {
+                        let _ = tx.send(Err(e));
+                        return;
+                    }
+                };
+
+                let metadata_manager = metadata_manager.clone();
+                let tx = tx.clone();
+                task_batch.push(tokio::spawn(async move {
+                    let table_info_values = match metadata_manager
+                        .table_info_manager()
+                        .batch_get(&table_ids)
+                        .await
+                        .context(TableMetadataManagerSnafu)
+                    {
+                        Ok(table_info_values) => table_info_values,
+                        Err(e) => {
+                            let _ = tx.send(Err(e));
+                            return;
+                        }
+                    };
+
+                    for table in table_info_values.into_values().map(build_table) {
+                        if tx.send(table).is_err() {
+                            return;
+                        }
+                    }
+                }));
+                if task_batch.len() == CONCURRENCY {
+                    let _ = futures::future::join_all(std::mem::take(&mut task_batch)).await;
                 }
             }
+            if !task_batch.is_empty() {
+                let _ = futures::future::join_all(std::mem::take(&mut task_batch)).await;
+            }
```

Comment:
How about using [Semaphore](https://docs.rs/tokio/latest/tokio/sync/struct.Semaphore.html#)?

Reply:
> How about using [Semaphore](https://docs.rs/tokio/latest/tokio/sync/struct.Semaphore.html#)?

fixed in 1d55a5f3cdd0e3d6bea62cf5948a91bafb8e7c65